firebase-firestore/src/main/java/com/google/firebase/firestore/model/protovalue/ObjectValue.java

```diff
@@ -0,0 +1,266 @@
+          overlayMap.subMap(path, createSuccessor(path)).keySet().iterator();
+      while (iterator.hasNext()) {
+        iterator.next();
+        iterator.remove();
+      }
+    }
+
+    /**
+     * Adds `value` to the overlay map at `path`. MapValues are recursively expanded into one
+     * overlay per leaf node.
+     */
+    private void setOverlay(FieldPath path, @Nullable Value value) {
+      if (!isType(value, TYPE_ORDER_OBJECT) || value.getMapValue().getFieldsCount() == 0) {
+        overlayMap.put(path, value);
+      } else {
+        for (Map.Entry<String, Value> entry : value.getMapValue().getFieldsMap().entrySet()) {
+          setOverlay(path.append(entry.getKey()), entry.getValue());
+        }
+      }
+    }
+
+    /** Returns an ObjectValue with all mutations applied. */
+    public ObjectValue build() {
+      if (overlayMap.isEmpty()) {
+        return baseObject;
+      } else {
+        MapValue.Builder result = baseObject.internalValue.getMapValue().toBuilder();
+        applyOverlay(FieldPath.EMPTY_PATH, result);
+        return new ObjectValue(Value.newBuilder().setMapValue(result).build());
+      }
+    }
+
+    /**
+     * Applies any overlays from `overlayMap` that exist at `currentPath` to the `resultAtPath` map.
+     * Overlays are expanded recursively based on their location in the backing ObjectValue's
+     * subtree and are processed by nesting level.
+     *
+     * <p>Example: Overlays { 'a.b.c' : 'foo', 'a.b.d' : 'bar', 'a.e' : 'foobar' }
+     *
+     * <p>To apply these overlays, the methods first creates a MapValue.Builder for `a`. It then
+     * calls applyOverlay() with a current path of `a` and the newly created MapValue.Builder. In
+     * its second call, `applyOverlay` assigns `a.b` to a new MapBuilder and `a.e` to 'foobar'. The
+     * third call assigns `a.b.c` and `a.b.d` to the MapValue.Builder created in the second step.
+     *
+     * <p>The overall aim of this method is to minimize conversions between MapValues and their
+     * builders.
+     *
+     * @param currentPath The path at the current nesting level. Can be set toFieldValue.EMPTY_PATH
+     *     to represent the root.
+     * @param resultAtPath A mutable copy of the existing data at the current nesting level.
+     *     Overlays are applied to this argument.
+     * @return Whether any modifications were applied (in any part of the subtree under
+     *     currentPath).
+     */
+    private boolean applyOverlay(FieldPath currentPath, MapValue.Builder resultAtPath) {
+      // Extract the data that exists at or below the current path. Te extracted subtree is
+      // subdivided during each iteration. The iteration stops when the slice becomes empty.
+      SortedMap<FieldPath, Value> currentSlice =
+          currentPath.isEmpty()
+              ? overlayMap
+              : overlayMap.subMap(currentPath, createSuccessor(currentPath));
+
+      boolean modified = false;
+
+      while (!currentSlice.isEmpty()) {
```

Comment:
I mainly used the TreeMap representation because it allowed me to store a `<FieldPath, Value>` map, which is more typesafe than any nested data structure I could come up with. If I trade type safety for readable code, the end result looks pretty much like you suggested. I am now storing the overlays in a `<String, Object>` map, where the key represents a single field segment and the value represents either:
- A `Value` proto for a leaf node (this includes MapValues that were set as part of a single user operation)
- A `Map<String, Value>` value that indicates additional nesting
- `null` to indicate field removals

This drastically simplifies the code  and reduce the runtime of `applyOverlay() ` to  `O(n * m)` (where `n` is the number of entries and `m` the nesting level). I believe before it was `O(n log n * m)`.